File: bindings/go/benchmarks/throughput.go
// Command throughput measures what crossing the boundary costs in the
// wickra-backtest Go binding.
//
// Every reach in this repository runs the same Rust engine, so a difference
// between two bindings is not a difference in the backtester -- it is the price
// of that language's FFI, paid once per bar on the streaming path and once per
// run on the batch path. That is the number worth knowing before choosing where
// to drive a live loop from.
//
// The strategy is examples/ema-cross.json: two EMAs, a crossover, fractional
// sizing, taker costs, slippage and a trailing stop. A realistic bar rather
// than an empty one, so the figure includes the engine work a real strategy
// does.
//
// Build the C ABI first and stage it under bindings/go/lib/<goos>_<goarch>/,
// as the binding's README describes, then:
//
//	cd bindings/go && go run ./benchmarks                  # 200k bars
//	cd bindings/go && go run ./benchmarks --bars 1000000
package main

import (
	"flag"
	"fmt"
	"log"
	"math"
	"sort"
	"time"

	wbt "github.com/wickra-lib/wickra-backtest-go"
)

const spec = `{"symbol":"BTCUSDT","timeframe":"1h",` +
	`"indicators":{"ema_fast":{"type":"Ema","params":[5]},` +
	`"ema_slow":{"type":"Ema","params":[15]}},` +
	`"entry":{"cross_above":["ema_fast","ema_slow"]},` +
	`"exit":{"cross_below":["ema_fast","ema_slow"]},` +
	`"sizing":{"type":"fixed_fraction","fraction":0.95},` +
	`"costs":{"taker_bps":5,"slippage":{"type":"fixed_bps","bps":2}},` +
	`"risk":{"trailing_stop_pct":5.0}}`

const capital = 10000.0

// series builds the deterministic synthetic OHLCV every binding's harness uses.
// No RNG, so two runs are comparable and so are two languages.
func series(bars int) (open, high, low, closes, volume []float64, times []int64) {
	open = make([]float64, bars)
	high = make([]float64, bars)
	low = make([]float64, bars)
	closes = make([]float64, bars)
	volume = make([]float64, bars)
	times = make([]int64, bars)
	for i := 0; i < bars; i++ {
		mid := 100 + math.Sin(float64(i)*0.001)*20 + float64(i)*1e-4
		closes[i] = mid + math.Sin(float64(i)*0.05)*2
		if i == 0 {
			open[i] = closes[i]
		} else {
			open[i] = closes[i-1]
		}
		high[i] = math.Max(open[i], closes[i]) + 1.5
		low[i] = math.Min(open[i], closes[i]) - 1.5
		volume[i] = 1000 + float64(i%97)*13
		times[i] = int64(i)
	}
	return
}

// medianDuration runs fn once to warm up, then reports the median of reps runs.
func medianDuration(fn func(), reps int) time.Duration {
	fn()
	samples := make([]time.Duration, reps)
	for i := range samples {
		start := time.Now()
		fn()
		samples[i] = time.Since(start)
	}
	sort.Slice(samples, func(a, b int) bool { return samples[a] < samples[b] })
	return samples[len(samples)/2]
}

func main() {
	bars := flag.Int("bars", 200000, "bars per run")
	flag.Parse()
	if *bars < 1000 {
		log.Fatal("--bars must be at least 1000")
	}

	open, high, low, closes, volume, times := series(*bars)

	streaming := medianDuration(func() {
		live, err := wbt.NewStreamingBacktest(spec, capital)
		if err != nil {
			log.Fatalf("could not start the run: %v", err)
		}
		defer live.Close()
		for i := 0; i < *bars; i++ {
			if err := live.Step(open[i], high[i], low[i], closes[i], volume[i], times[i]); err != nil {
				log.Fatalf("bar %d rejected: %v", i, err)
			}
		}
		if _, err := live.FinishJSON(); err != nil {
			log.Fatalf("could not finish the run: %v", err)
		}
	}, 3)

	batch := medianDuration(func() {
		if _, err := wbt.Run(open, high, low, closes, volume, times, spec, capital); err != nil {
			log.Fatalf("batch run failed: %v", err)
		}
	}, 3)

	fmt.Printf("wickra-backtest Go throughput — %s bars (median of 3 runs)\n\n", commas(int64(*bars)))
	fmt.Printf("%-14s%16s%12s\n", "path", "bars/sec", "ns/bar")
	fmt.Println("------------------------------------------")
	for _, row := range []struct {
		name string
		took time.Duration
	}{{"streaming", streaming}, {"batch", batch}} {
		perSecond := float64(*bars) / row.took.Seconds()
		nsPerBar := float64(row.took.Nanoseconds()) / float64(*bars)
		fmt.Printf("%-14s%16s%12s\n", row.name, commas(int64(perSecond)), commas(int64(nsPerBar)))
	}
	fmt.Print("\nStreaming crosses the boundary once per bar, with scalars; batch crosses it\n",
		"once per run and hands over six slices. cgo charges per call, so which of\n",
		"the two wins is a property of the language, not of the engine behind both.\n",
		"Machine-dependent — compare bindings on one machine, not across machines.\n")
}

// commas renders n with thousands separators, which the standard library does
// not offer and which one extra dependency would not justify.
func commas(n int64) string {
	digits := fmt.Sprintf("%d", n)
	sign := ""
	if digits[0] == '-' {
		sign, digits = "-", digits[1:]
	}
	out := make([]byte, 0, len(digits)+len(digits)/3)
	for i, c := range []byte(digits) {
		if i > 0 && (len(digits)-i)%3 == 0 {
			out = append(out, ',')
		}
		out = append(out, c)
	}
	return sign + string(out)
}

File: bindings/r/benchmarks/throughput.R
# Throughput benchmark for the wickra-backtest R binding.
#
# Measures what crossing the boundary costs. Every reach in this repository runs
# the same Rust engine, so a difference between two bindings is not a difference
# in the backtester -- it is the price of that language's FFI, paid once per bar
# on the streaming path and once per run on the batch path.
#
# R is the boundary where that difference is largest and most worth knowing: the
# streaming loop is an interpreted `for` over .Call, while the batch path hands
# over six numeric vectors and returns once.
#
# The strategy is examples/ema-cross.json: two EMAs, a crossover, fractional
# sizing, taker costs, slippage and a trailing stop.
#
# Run after installing the binding (R CMD INSTALL bindings/r):
#
#   Rscript bindings/r/benchmarks/throughput.R                 # 200k bars
#   Rscript bindings/r/benchmarks/throughput.R --bars 1000000

library(wickrabacktest)

spec <- paste0(
  '{"symbol":"BTCUSDT","timeframe":"1h",',
  '"indicators":{"ema_fast":{"type":"Ema","params":[5]},',
  '"ema_slow":{"type":"Ema","params":[15]}},',
  '"entry":{"cross_above":["ema_fast","ema_slow"]},',
  '"exit":{"cross_below":["ema_fast","ema_slow"]},',
  '"sizing":{"type":"fixed_fraction","fraction":0.95},',
  '"costs":{"taker_bps":5,"slippage":{"type":"fixed_bps","bps":2}},',
  '"risk":{"trailing_stop_pct":5.0}}'
)
capital <- 10000

parse_bars <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  hit <- match("--bars", args)
  if (is.na(hit) || length(args) < hit + 1L) {
    return(200000L)
  }
  n <- suppressWarnings(as.integer(args[hit + 1L]))
  if (is.na(n) || n < 1000L) {
    stop("--bars must be an integer >= 1000")
  }
  n
}

bars <- parse_bars()

# Deterministic synthetic OHLCV, from the same formula every binding's harness
# uses. Vectorised, so building it does not dominate the measurement.
i <- seq_len(bars) - 1L
mid <- 100 + sin(i * 0.001) * 20 + i * 1e-4
close <- mid + sin(i * 0.05) * 2
open <- c(close[1], close[-bars])
high <- pmax(open, close) + 1.5
low <- pmin(open, close) - 1.5
volume <- 1000 + (i %% 97) * 13
time <- as.numeric(i)

# Seconds for one run, measured by repeating until the clock has something to
# say. R's timers are coarse on some platforms -- a fast batch run over a small
# series measured as exactly zero and reported an infinite rate -- so the run is
# repeated until the batch takes at least a fifth of a second, and the total is
# divided by the count.
median_seconds <- function(fn, reps = 3L, floor_seconds = 0.2) {
  fn() # warmup
  samples <- vapply(seq_len(reps), function(...) {
    count <- 1L
    repeat {
      start <- Sys.time()
      for (unused in seq_len(count)) fn()
      elapsed <- as.numeric(Sys.time() - start, units = "secs")
      if (elapsed >= floor_seconds) {
        return(elapsed / count)
      }
      count <- count * 2L
    }
  }, numeric(1))
  stats::median(samples)
}

streaming <- function() {
  bt <- backtest_stream_new(spec, capital = capital)
  for (k in seq_len(bars)) {
    backtest_stream_step(bt, open[k], high[k], low[k], close[k], volume[k], time[k])
  }
  invisible(backtest_stream_finish_json(bt))
}

batch <- function() {
  invisible(backtest_run(open, high, low, close,
                         volume = volume, time = time,
                         spec = spec, capital = capital))
}

streaming_s <- median_seconds(streaming)
batch_s <- median_seconds(batch)

cat(sprintf("wickra-backtest R throughput — %s bars (median of 3 runs)\n\n",
            format(bars, big.mark = ",")))
cat(sprintf("%-14s%16s%12s\n", "path", "bars/sec", "ns/bar"))
cat(strrep("-", 42), "\n", sep = "")
for (row in list(list("streaming", streaming_s), list("batch", batch_s))) {
  cat(sprintf("%-14s%16s%12s\n", row[[1]],
              format(round(bars / row[[2]]), big.mark = ","),
              format(round(row[[2]] / bars * 1e9), big.mark = ",")))
}
cat(paste0(
  "\nStreaming crosses the boundary once per bar, from an interpreted loop.\n",
  "Batch crosses it once per run and hands over six numeric vectors, so the\n",
  "gap here is R's per-call overhead rather than anything about the engine.\n",
  "Machine-dependent — compare bindings on one machine, not across machines.\n"
))
